jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java

```diff
@@ -177,47 +115,60 @@ public static String resolveMainClass(
   }
 
   /**
-   * Searches for a .class file containing a main method in a root directory.
+   * Finds the classes with {@code public static void main(String[] args)} in {@code rootDirectory}.
    *
-   * @return the name of the class if one is found, null if no class is found.
-   * @throws IOException if searching/reading files fails.
+   * @param rootDirectory directory containing the {@code .class} files
    */
   @VisibleForTesting
   static List<String> findMainClasses(Path rootDirectory) throws IOException {
-    List<String> classNames = new ArrayList<>();
-
-    // Make sure rootDirectory is valid
+    // Makes sure rootDirectory is valid.
     if (!Files.exists(rootDirectory) || !Files.isDirectory(rootDirectory)) {
-      return classNames;
+      return Collections.emptyList();
     }
 
-    // Get all .class files
-    ClassFileLoader classFileLoader = new ClassFileLoader(rootDirectory);
-    new DirectoryWalker(rootDirectory)
-        .filter(Files::isRegularFile)
-        .filter(path -> path.toString().endsWith(".class"))
-        .walk(
-            classFile -> {
-              Class<?> fileClass = classFileLoader.findClass(null, classFile);
-              if (fileClass == null) {
-                return;
-              }
-              try {
-                // Check if class contains {@code public static void main(String[] args)}
-                Method main = fileClass.getMethod("main", String[].class);
-                if (main != null
-                    && main.getReturnType() == void.class
-                    && Modifier.isStatic(main.getModifiers())
-                    && Modifier.isPublic(main.getModifiers())) {
-                  classNames.add(fileClass.getName());
+    List<String> classNames = new ArrayList<>();
+
+    ClassPool classPool = new ClassPool();
+    classPool.appendSystemPath();
+
+    try {
+      CtClass[] mainMethodParams = new CtClass[] {classPool.get("java.lang.String[]")};
+
+      new DirectoryWalker(rootDirectory)
+          .filter(Files::isRegularFile)
+          .filter(path -> path.toString().endsWith(".class"))
+          .walk(
+              classFile -> {
+                try (InputStream classFileInputStream = Files.newInputStream(classFile)) {
+                  CtClass fileClass = classPool.makeClass(classFileInputStream);
+
+                  try {
```

Comment:
Would it be cleaner to have just one try block with multiple catch clauses instead of having them nested?